builder/build.go

```diff
@@ -0,0 +1,1197 @@
+		return nil, fmt.Errorf("error opening SquashFS layer: %s", err)
+	}
+	h := sha512.New512_256()
+	dst, err := os.Create(b.layerPath(id))
+	if err != nil {
+		return nil, fmt.Errorf("error writing to layer cache: %s", err)
+	}
+	defer dst.Close()
+	if _, err := io.Copy(dst, io.TeeReader(f, h)); err != nil {
+		return nil, fmt.Errorf("error writing to layer cache: %s", err)
+	}
+	layer = &ct.ImageLayer{
+		ID:     id,
+		Type:   ct.ImageLayerTypeSquashfs,
+		Length: stat.Size(),
+		Hashes: map[string]string{
+			"sha512_256": hex.EncodeToString(h.Sum(nil)),
+		},
+	}
+	data, err := json.Marshal(layer)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding layer config: %s", err)
+	}
+	if err := ioutil.WriteFile(b.layerConfigPath(id), data, 0644); err != nil {
+		return nil, fmt.Errorf("error writing to layer cache: %s", err)
+	}
+	return layer, nil
+}
+
+func (b *Builder) GoInputsFor(platform GoPlatform) *GoInputs {
+	b.goInputsMtx.Lock()
+	defer b.goInputsMtx.Unlock()
+	if g, ok := b.goInputs[platform]; ok {
+		return g
+	}
+	g := NewGoInputs(platform)
+	b.goInputs[platform] = g
+	return g
+}
+
+const layerURLTemplate = "file:///var/lib/flynn/layer-cache/{id}.squashfs"
+
+func (b *Builder) layerPath(id string) string {
+	return fmt.Sprintf("/var/lib/flynn/layer-cache/%s.squashfs", id)
+}
+
+func (b *Builder) layerConfigPath(id string) string {
+	return fmt.Sprintf("/var/lib/flynn/layer-cache/%s.json", id)
+}
+
+// generateLayerID generates a layer ID from a set of all inputs required to
+// build the layer, which prevents rebuilding a layer if the inputs haven't
+// changed.
+func (b *Builder) generateLayerID(name string, run []string, env map[string]string, artifact *ct.Artifact, inputs ...string) (id string, err error) {
+	start := time.Now()
+	defer func() {
+		b.log.Debug("generated layer ID", "name", name, "id", id, "duration", time.Since(start))
+	}()
+
+	h := sha512.New512_256()
+
+	// hash on the layer name
+	h.Write([]byte(name))
+
+	// hash on the list of commands being run
+	if len(run) > 0 {
+		h.Write([]byte(strings.Join(run, "")))
+	}
+
+	// hash on the sorted list of environment variables
+	if len(env) > 0 {
+		sortEnv := make([]string, 0, len(env))
+		for k, v := range env {
+			sortEnv = append(sortEnv, fmt.Sprintf("%s=%s", k, v))
+		}
+		sort.Strings(sortEnv)
+		h.Write([]byte(strings.Join(sortEnv, "")))
+	}
+
+	// hash on the artifact manifest
+	if artifact != nil {
+		h.Write(artifact.RawManifest)
+	}
+
+	// hash on the filename, size and content of each input
+	//
+	// TODO: consider storing a map of filenames to hashes and cache based
+	//       on the last modified time to reduce build time
+	for _, input := range inputs {
+		f, err := os.Open(input)
+		if err != nil {
+			return "", err
+		}
+		stat, err := f.Stat()
+		if err != nil {
+			f.Close()
+			return "", err
+		}
+		fmt.Fprintf(h, "%s%d", input, stat.Size())
+		_, err = io.Copy(h, f)
+		f.Close()
+		if err != nil {
+			return "", err
+		}
+	}
```

Comment:
I think all of the details that are written to the hash should be encoded to a JSON object which is subsequently hashed. This will completely avoid a few (probably extremely rare) edge cases that could be caused by for example joining command line arguments together without a delimiter.